shotover/src/transforms/kafka/sink_cluster/mod.rs

```diff
@@ -2738,180 +2764,229 @@ impl KafkaSinkCluster {
             // In fact the java driver doesnt even support multiple coordinators of different types yet:
             // https://github.com/apache/kafka/blob/4825c89d14e5f1b2da7e1f48dac97888602028d7/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L921
             //
-            // So, just like with version <= 3, we just pick the first shotover node in the rack of the coordinator for each coordinator.
+            // So, just like with version <= 3, we try to deterministically pick one UP shotover node in the rack of the coordinator for each coordinator.
+            // If there is no UP shotover node in the rack, we try to deterministically pick one UP shotover node out of the rack of the coordinator.
+            // If there is no UP shotover node at all, return the COORDINATOR_NOT_AVAILABLE error.
             for coordinator in &mut find_coordinator.coordinators {
                 // skip rewriting on error
                 if coordinator.error_code == 0 {
+                    let hash = hash_str_bytes(coordinator.key.clone());
                     let coordinator_rack = &self
                         .nodes
                         .iter()
                         .find(|x| x.broker_id == coordinator.node_id)
                         .unwrap()
                         .rack
                         .as_ref();
-                    let shotover_node = self
-                        .shotover_nodes
-                        .iter()
-                        .find(|shotover_node| {
-                            coordinator_rack
-                                .map(|rack| rack == &shotover_node.rack)
-                                .unwrap_or(true)
-                        })
-                        .unwrap();
-                    coordinator.host = shotover_node.address.host.clone();
-                    coordinator.port = shotover_node.address.port;
-                    coordinator.node_id = shotover_node.broker_id;
+
+                    let (shotover_nodes_in_rack, shotover_nodes_out_of_rack): (
+                        Vec<&ShotoverNode>,
+                        Vec<&ShotoverNode>,
+                    ) = up_shotover_nodes.iter().partition(|shotover_node| {
+                        coordinator_rack
+                            .map(|rack| rack == &shotover_node.rack)
+                            .unwrap_or(true)
+                    });
+
+                    let shotover_node: Option<&&ShotoverNode> = shotover_nodes_in_rack
+                        .get(hash % shotover_nodes_in_rack.len())
+                        .or_else(|| {
+                            shotover_nodes_out_of_rack.get(hash % shotover_nodes_out_of_rack.len())
+                        });
+
+                    match shotover_node {
+                        Some(shotover_node) => {
+                            find_coordinator.host = shotover_node.address.host.clone();
```

Comment:
here is the encoding problem, we are setting find_coordinator.host/port/node_id instead of coordinator.host/port/node_id

Reply:
fixed